### PR TITLE
[Pre-work] Prepare @tracked decorator

### PR DESCRIPTION
**Pull request by chadhietala**
This is pre-work to expose `tracked` as a decorator. In this PR @smfoote moved the tests over from being a helper function that augmented the descriptors to an actual function that can be used as a decorator. This also updates all the tests to test `@tracked` in static form instead of property descriptor helpers.

/cc @wycats @tomdale

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (btecu) [inline .vscode/settings.json]: This file shouldn't be commited.

File: packages/ember-metal/lib/tracked.ts [hunk 2]
```diff
@@ -96,12 +96,13 @@ class Tracker {
...
   @param dependencies Optional dependents to be tracked.
  */
-export function tracked(
-  _target: object,
-  key: string,
-  descriptor: PropertyDescriptor
-): PropertyDescriptor {
-  if ('value' in descriptor) {
+export function tracked(...dependencies: string[]): MethodDecorator;
+export function tracked(target: any, key: any): any;
```
Comment (tomdale) [inline packages/ember-metal/lib/tracked.ts]: I believe these can be typed as `unknown` and `PropertyKey`, respectively.

File: packages/ember-metal/tests/tracked/get_test.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        let obj = new Obj();
+
+        this.assert.equal(get(obj, ''), 'empty');
+      }
+
+      '@test should get a @tracked path'() {
+        class Key {
+          @tracked value = 'value';
+          constructor() {
+            this.value = 'value';
```
Comment (tomdale) [inline packages/ember-metal/tests/tracked/get_test.ts]: Why do we need to set this in the constructor if we're already initializing the property above?

File: packages/ember-metal/tests/tracked/set_test.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    return new Obj();
+  };
+
+  moduleFor(
+    '@tracked set',
+    class extends AbstractTestCase {
+      teardown() {
+        setHasViews(() => false);
```
Comment (tomdale) [inline packages/ember-metal/tests/tracked/set_test.ts]: Why is this needed?

File: packages/ember-metal/tests/tracked/validation_test.ts [hunk 5]
```diff
@@ -85,32 +89,35 @@ if (EMBER_METAL_TRACKED_PROPERTIES) {
...
+            self = this;
           }
         }
 
         defineProperty(
           EmberObject.prototype,
           'full',
           computed('name', function() {
-            let name = get(this, 'name');
+            let name = get(self, 'name');
```
Comment (tomdale) [inline packages/ember-metal/tests/tracked/validation_test.ts]: I think you're having to introduce the `self` local variable because TypeScript can't infer the type of `this` inside this computed property. If that's the case, a better way to do this would be to annotate the `this` type:

```ts
computed('name', function(this: EmberObject) { /*...*/ })
```

**Review by tomdale — CHANGES_REQUESTED** [review]
Looks great, thank you for doing this. Just a few nitpicks, and I'd like to understand why in the tests we're initializing properties and then re-initializing them in the constructor to the same value. Unless I'm missing something, they're semantically identical.

File: packages/ember-metal/tests/tracked/get_test.ts (hunk 3 above)
Comment (smfoote) [reply packages/ember-metal/tests/tracked/get_test.ts]: Fixed

File: packages/ember-metal/lib/tracked.ts (hunk 2 above)
Comment (smfoote) [reply packages/ember-metal/lib/tracked.ts]: Fixed

File: packages/ember-metal/tests/tracked/set_test.ts (hunk 4 above)
Comment (smfoote) [reply packages/ember-metal/tests/tracked/set_test.ts]: Apparently it's not. Deleted.

File: packages/ember-metal/tests/tracked/validation_test.ts (hunk 5 above)
Comment (smfoote) [reply packages/ember-metal/tests/tracked/validation_test.ts]: Cool! Done.
